### ISSUE TITLE
Threading layer: Race condition in the layer causes incorrect "object simultaneously used in thread .. and thread .." errors to be reported

### ISSUE THREAD

**Issue by DominikWitczakAMD** [KhronosGroup/Vulkan-LoaderAndValidationLayers#2607]
Consider the following scenario:

1. Thread A calls vkDestroyBuffer()
2. Thread B calls vkCreateBuffer()

It is legitimate for the driver to return the same handle if the calls are executed in a serialized manner, in order: A followed by B.

The validation layer chain does some additional chores after VK impl returns execution flow from vkDestroyBuffer(), before passing control back to the app. However, this additional housekeeping work does not seem to be protected by any sort of a critical section. This can lead to a situation where the threading layer notices the same handle is being used in two threads at the same time, even though that very handle has already had ownership released from the original object and passed over to a new object.

**Comment by DominikWitczakAMD** [comment]
I have a repro for this if that would be helpful.

**Comment by DominikWitczakAMD** [comment]
Forgot to mention: if the error is ignored, the layer first does an assert(0) and then crashes afterward due to a NPE.

**Comment by chrisforbes** [comment]
How are you stacking up the layers for this? CV etc relies on unique_objects to guarantee that we don't see reuse of nondispatchable handles.

**Comment by DominikWitczakAMD** [comment]
Fair point. IIRC, I used alphabetic ordering when typing all the custom paths to json files in the registry so that would mean unique_objects is loaded before core_validation. I'll get back to you on this tomorrow, I'm OoO already.

As a side point, are you considering adding some sort of a requirements chain to the layers? Issues like these are pretty hardcore to nail down if you're not knee-deep into VL project's guts :-/ A warning along the lines of "layer X should come before layer Y or else!" would suffice just fine.

**Comment by chrisforbes** [comment]
You should already be getting such a warning, at least for CV misordered wrt UO.

The standard_validation pseudo-layer enables the layers in the order that is known to work.
